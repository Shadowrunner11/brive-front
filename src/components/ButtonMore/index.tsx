
import { MouseEvent, useCallback } from 'react'
import { cva } from 'class-variance-authority'

import styles from './styles.module.css'

const buttonMore = cva(styles.button)

export const ButtonMore = () => {
  const _handleClick = useCallback((event: MouseEvent<HTMLButtonElement>)=>{
    event.preventDefault()
  }, [])

  return <button onClick={_handleClick} className={buttonMore()}>
    More
  </button>
}
