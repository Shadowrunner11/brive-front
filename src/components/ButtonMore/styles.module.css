/* TODO: Separate structure form theme to use CVA */

.button{
  background-color: cadetblue;
  display: block;
  width: 100%;
  border-radius: .5rem .5rem 1rem 1rem;
}

.button:hover{
  background-color: rgb(8, 180, 168);
  cursor: pointer;
}
