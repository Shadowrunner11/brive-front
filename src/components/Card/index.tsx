import blur from '../../assets/blur.jpeg'

import { LazyLoadImage } from "react-lazy-load-image-component";
import { cva, type VariantProps } from "class-variance-authority";

import styles from './styles.module.css'


const card = cva(styles.card, {
  variants:{
    intent:{
      primary: styles.card_themePrimary
    }
  }
})

interface CardProps extends VariantProps<typeof card>  {
  cardMediaURL?: string | null;
  cardContent: React.ReactNode;
  cardAction?: JSX.Element;
  cardTop?: React.ReactNode
}


export const Card = ({ cardMediaURL, cardAction, cardContent, cardTop, intent = 'primary'}: CardProps)=> (
  <div className={card({intent})}>
    {
      cardTop && <div>
        {cardTop}
      </div>
    }

    <div className={styles.card__media}>
      {
        cardMediaURL && 
          <LazyLoadImage placeholderSrc={blur} width='18rem' height={300} src={cardMediaURL} alt='card media'/>
      }
    </div>

    {cardContent}

    {
      cardAction &&
        <div className={styles.card__cardAction}>
          {cardAction}
        </div>
    }

  </div>
)
