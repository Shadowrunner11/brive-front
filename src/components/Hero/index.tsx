import styles from './styles.module.css'

export const Hero = ()=> (
  <div className={styles.hero} >
    <div className={styles.hero__text}>
      <h1><strong>Rick and Morty</strong></h1>
      <h2><strong>Se all characters. <br /> And more.</strong></h2>
    </div>
  </div>
)
