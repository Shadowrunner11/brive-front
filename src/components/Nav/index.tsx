import styles from "./styles.module.css";
import logo from "../../assets/Rick_and_Morty.svg";

export const Nav = () => (
  <header className={styles.header}>
    <div className={styles.header__logo}>
      <img src={logo} alt="Rick and Morty logo" />
    </div>
    <p>
      <strong>Terms + Conditions</strong>
    </p>
  </header>
);
