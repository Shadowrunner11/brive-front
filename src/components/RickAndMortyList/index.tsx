import { useInfiniteQuery } from "@tanstack/react-query";
import { GridItemProps, VirtuosoGrid } from "react-virtuoso";
import {
  PropsWithChildren,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";
import { Character } from "../../gql/graphql";
import { Card } from "../../components/Card";

import styles from "./styles.module.css";

import styled from "@emotion/styled";
import { ButtonMore } from "../ButtonMore";
import { grahqlClient as graphqlClient } from "../../lib/modules";

const ItemContentCard = (
  index: number,
  { image, name, species }: Character,
) => (
  <Card
    key={index + (name ?? "no content")}
    cardTop={
      <h3>
        <strong>{name}</strong>
      </h3>
    }
    cardMediaURL={image}
    cardAction={<ButtonMore />}
    cardContent={
      <div>
        <h3>{species}</h3>
      </div>
    }
  />
);

const Footer = () => <div>Loading...</div>;

const Item = ({ children }: PropsWithChildren<GridItemProps>) => (
  <div className={styles.item}>{children}</div>
);

// TODO should not depend on styled components, not a portable solution
const List: any = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 2rem;
  justify-content: center;
`;

export const RickAndMortyList = () => {
  // TODO: does this really need to be a state? could be a ref
  const [page, setPage] = useState(1);

  // TODO: check if IoC is possible
  const { data, fetchNextPage } = useInfiniteQuery(
    ["rickAndMorty"],
    async ({ pageParam }) =>
      await graphqlClient.getPaginatedCharacters(pageParam),
  );

  // TODO: optimize with pseudo useSingal and memoized pattern
  const flattedData = useMemo(
    () => data?.pages?.flatMap((character) => character?.results ?? []),
    [data?.pages],
  );

  const fetchMore = useCallback(() => {
    setPage((prevPage) => prevPage + 1);
  }, []);

  useEffect(() => {
    fetchNextPage({
      pageParam: page,
    });
  }, [page, fetchNextPage]);

  if (!flattedData?.length) return <Footer />;

  return (
    <VirtuosoGrid
      useWindowScroll
      className={styles.virtuoso}
      style={{ height: 400 }}
      overscan={20}
      endReached={fetchMore}
      components={{
        Item,
        List,
        Footer,
      }}
      data={(flattedData as Character[]) ?? []}
      itemContent={ItemContentCard}
    />
  );
};
