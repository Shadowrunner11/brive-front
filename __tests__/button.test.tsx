import renderer from "react-test-renderer";
import { describe, it, expect } from "@jest/globals";

import { ButtonMore } from "../src/components/ButtonMore";

describe("Card tests", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<ButtonMore />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
