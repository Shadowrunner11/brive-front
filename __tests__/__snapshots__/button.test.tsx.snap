// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card tests renders correctly 1`] = `
<button
  className="asdsad"
  onClick={[Function]}
>
  More
</button>
`;
